### ISSUE TITLE
Update HTTPS certificate to support *.*.altis.dev

### ISSUE THREAD

**Issue by roborourke** [humanmade/altis-local-server#135]
One of the limitations we have with local server is the inability to support sub-sub domains of `altis.dev`. This makes it hard to migrate data from production or to test out sites that employ logic around having subdomains or the `SUBDOMAIN_INSTALL` constant.

Related to #126 

Acceptance criteria:

- [ ] sub-subdomains of `altis.dev` are supported and served over https automatically

**Comment by roborourke** [comment]
Marking this as future release because it's literally impossible with a single certificate. This would require significant amount of work to change to either generating self signed certs and adding those programmatically to a trusted store as per local-chassis or building out some sort of service that updates the `altis.dev` DNS records on demand for provisioning certs with LetsEncrypt.

**Comment by goldenapples** [comment]
I'm running up against this right now, and would love to see one of the following: 
- instructions for generating a self-signed certificate and configuring it to be used within altis-local-server-nginx. 
- the ability to run local-server with HSTS disabled.

**Comment by mikeselander** [comment]
I'm also now running against this on a separate client from Than. Both of his requests would be exceedingly helpful if we can't setup the certificates correctly automatically.

**Comment by nathanielks** [comment]
I don't have time to read this now, but i want to leave this for my future self: https://www.smarthomebeginner.com/traefik-2-docker-tutorial/

**Comment by nathanielks** [comment]
Also making note that Caddy 2 might be a good replacement as it can manage/generate TLS certificates: https://caddyserver.com/v2

**Comment by nathanielks** [comment]
TL;DR: they generate their own trusted CA locally: https://caddyserver.com/docs/automatic-https#local-https

**Comment by rmccue** [comment]
Caddy's is powered by [smallstep](https://github.com/smallstep/certificates) under the hood; related Chassis discussion: https://github.com/Chassis/chassis_openssl/issues/15#issuecomment-615474478

**Comment by jennybeaumont** [comment]
Closing in favor of https://github.com/humanmade/altis-local-server/issues/151

**Comment by roborourke** [comment]
For anyone googling this - a short term workaround if you're using Google Chrome is to type `thisisunsafe` on the HTTPS warning screen you get when navigating to a sub-sub domain of `altis.dev`

**Comment by nathanielks** [comment]
Without context, I don't see #151 solving this issue. Is there more context that can be added?

**Comment by jennybeaumont** [comment]
It was determined that this isn't technically possible.

**Comment by nathanielks** [comment]
FWIW I do think this is technically possible. When spinning up local-server, we know what the second star is in the url (`*.<project name>.altis.dev`). Now it's a matter of enabling SSL certificate creation when the project is started up creating a wildcard certificate for `*.<project name>.altis.dev`.

**Comment by roborourke** [comment]
#151 hints at being able to choose the URL local server runs on, i.e. moving away from the `altis.dev` requirement. Maybe technically possible but very difficult and not something there's really been a strong will for anyone to dive into so far. If we can make SSL work `*.<project>.altis.dev` then presumably we can use that same solution for providing SSL for any desired local URL.

**Comment by nathanielks** [comment]
TL;DR: I've done some messing around this evening and this definitely appears possible with Traefik v2 and tools like [`mkcert`](https://github.com/FiloSottile/mkcert). Traefik v2 allows for dynamic configuration updates, meaning the proxy can be running and dynamically add new SSL certificates. Upgrading to Traefik v2 changes a good bit of how are configured, but I'd say it's within the realm of feasability, though requiring more than the 2 hours or so I've put looking into it 🙂

**Comment by nathanielks** [comment]
The other TL;DR is `composer local-server up` would create SSL certificates if they don't exist, place them in a folder Traefik has mounted as a volume, and manage a file Traefik would read for configuring itself with the SSL certificates.

**Comment by jennybeaumont** [comment]
Shall we reopen? Does this deliver value? Do we have a clear path forward?

**Comment by roborourke** [comment]
I think we more or less ruled out mkcert as it can’t be bundled with this package neatly. @rmccue mentioned using Caddy but I’ve not really managed to grok how to use it yet.

There is value in this I just thought we’d take a different approach. I’ll reopen.

**Comment by nathanielks** [comment]
I had similar issues with Caddy. I started with that first last night. It worked great for anything localhost/127.0.0.1, but DNS addresses it attempts to use Let's Encrypt to generate a certificate (even if the address resolves to localhost).

For mkcert, we could wrap the CLI like we do docker-compose, I reckon. I'm not dedicated to mkcert, either. I'm curious about Smallstep's tool, too, but didn't mess around with it at all.

**Comment by nathanielks** [comment]
All that to say, whenever this gets reprioritized, a spike to play around with what options we have available would be good 😀

**Comment by rmccue** [comment]
mkcert is not really programmatically usable, that's part of the reason I hadn't dug into it further for Chassis. And Smallstep is what's used inside Caddy I believe.

**Comment by joehoyle** [comment]
Not sure if it's mentioned here, but it's not possible to use self-signed certs on *.dev domains. It's also not possible to get a cert signed for `*.*.altis.dev` as far as I know (SSLMate were not able to, and said it's not supported by the spec)
